src/cmap/auth/mongodb_oidc/token_entry_cache.ts

```diff
@@ -0,0 +1,110 @@
+import type { OIDCMechanismServerStep1, OIDCRequestTokenResult } from '../mongodb_oidc';
+
+/* 5 minutes in milliseonds */
+const EXPIRATION_BUFFER = 300000;
+/* 5 hours in seconds */
+const DEFAULT_EXPIRATION = 18000;
+
+/** @internal */
+export class TokenEntry {
+  tokenResult: OIDCRequestTokenResult;
+  serverResult: OIDCMechanismServerStep1;
+  expiration: number;
+
+  /**
+   * Instantiate the entry.
+   */
+  constructor(
+    tokenResult: OIDCRequestTokenResult,
+    serverResult: OIDCMechanismServerStep1,
+    expiration: number
+  ) {
+    this.tokenResult = tokenResult;
+    this.serverResult = serverResult;
+    this.expiration = expiration;
+  }
+
+  /**
+   * The entry is still valid if the expiration is more than
+   * 5 minutes from the expiration time.
+   */
+  isValid() {
+    return this.expiration - Date.now() > EXPIRATION_BUFFER;
+  }
+}
+
+/**
+ * Cache of OIDC token entries.
+ * @internal
+ */
+export class TokenEntryCache {
+  entries: Map<string, TokenEntry>;
+
+  constructor() {
+    this.entries = new Map();
+  }
+
+  /**
+   * Set an entry in the token cache.
+   */
+  addEntry(
+    address: string,
+    username: string,
+    tokenResult: OIDCRequestTokenResult,
+    serverResult: OIDCMechanismServerStep1
+  ): TokenEntry {
+    const entry = new TokenEntry(
+      tokenResult,
+      serverResult,
+      expirationTime(tokenResult.expiresInSeconds)
```

Comment:
Just a note, from the spec:

> If there is no `expiresInSeconds`, the token must be consider expired after first usage. 

`expirationTime(0)` is `Date.now()`, but there might be some question about how to interpret “no `expiresInSeconds`” in JS (e.g. do null and undefined fall into that category?).

Fwiw, from the devtools code we’re planning to always make use of this line and provide `expirationTime: 0` or whatever else allows us to bypass the global cache.

Reply:
I've added the units to the constants and decided to default `null` and `undefined` to equal `0` in this case.

Reply:
Where is the defaulting happening?

Reply:
https://github.com/mongodb/node-mongodb-native/pull/3577/files/6cfd1d5f36a9a9363e7c29b9bac3f37e32aac4ce#diff-bd6cb7449c5cb3124bcfb31066ba2c2374aea822f8d1326b4ba111e501105e04R102

Reply:
_if_ we're provided a `null` value (which is disallowed by TS - but maybe something we should consider anyways because this will be arbitrary dynamic user input.  It never hurts to be safe.  I mentioned here because you were talking about it), it only works because of the way JavaScript coerces numbers when performing multiplication.

```typescript
function foo(x = "default") {
	return x
}

foo(undefined) // "default"
foo() // "default"
foo(null) // null
```

This function works when provided null because `<number> + null * number` is always zero, because `null` will get coerced to `0`.

_if_ we're going to support a `null` return value here, I think we should make it explicit (remove the default value, make it optional and use `??` to provide a default value inside the function).  

_if_ we decide not to support `null` here, then I think it's fine as-is

Reply:
I've changed to set the default in the function now to support null getting passed in.